### PR TITLE
mempool: add `nop` mempool

### LINKED ISSUE #1666: `nop` mempool
ADR-111: https://github.com/cometbft/cometbft/pull/1585

## Summary

> The current mempool built into CometBFT implements a robust yet somewhat inefficient transaction gossip mechanism.
While the CometBFT team is currently working on more efficient general-purpose transaction gossiping mechanisms,
some users have expressed their desire to manage both the mempool and the transaction dissemination mechanism
outside CometBFT (typically at the application level). @sergio-mena 

## Solution

Introduce `mempool.type` config setting with `flood` and `nop` options.

 `flood` would represent the existing `CListMempool`. 

`nop`, short for no operation, would disable all mempool-related functionality in CometBFT. When `nop` is being used, the ABCI developers will need to design and implement the application side (or external) mempool and propose transactions via `PrepareProposal`.

When `nop` mempool is used:

- `broadcast_*` RPC endpoints will return an error
- no P2P reactor will be running (peers attempting to send mempool msgs will be stopped)
- The mempool ABCI connection will still be created (should we make it optional?)
- `RequestPrepareProposal` will contain 0 transactions
- `create_empty_blocks=false` will return an error upon startup

### PR DESCRIPTION
**Pull request by melekes**
Closes #1666 

See
[ADR-111](https://github.com/cometbft/cometbft/pull/1585)

---

#### PR checklist

- [x] Tests written/updated
- [x] Changelog entry added in `.changelog` (we use [unclog](https://github.com/informalsystems/unclog) to manage our changelog)
- [x] Updated relevant documentation (`docs/` or `spec/`) and code comments

### CODE REVIEW

**Review by hvanz — APPROVED** [review]
Looks great! We just need to remove the label to backport to v0.34, which doesn't make much sense.

File: docs/core/mempool.md [hunk 1]
```diff
@@ -4,7 +4,41 @@ order: 12
...
+room for transactions that have not yet been committed.
+
+CometBFT currently supports two types of mempools: `flood` and `nop`.
+
+## 1. Flood
+
+The `flood` mempool stores transactions in a concurrent linked list. When a new
+transaction is received, it first checks there's a space for it (`size` and
+`max_txs_bytes` config options) and that it's not too big (`max_tx_bytes` config
+option). Then, it checks if this transaction has already been seen before using
```
Comment (sergio-mena) [inline docs/core/mempool.md:18]:
```suggestion
option). Then, it checks if this transaction has already been seen before by using
```

File: docs/core/mempool.md [hunk 2]
```diff
@@ -4,7 +4,41 @@ order: 12
...
+an LRU cache (`cache_size` regulates the cache's size). If all checks pass and
+the transaction is not in the cache (meaning it's new), the ABCI
+[`CheckTxAsync`][1] method is called. The ABCI application validates the
+transaction using its own rules.
+
+If the transaction is deemed valid by the ABCI application, it's added to the linked list.
+
+The mempools name (`flood`) comes from the dissemination mechanism. When a new
```
Comment (sergio-mena) [inline docs/core/mempool.md:26]:
```suggestion
The mempool's name (`flood`) comes from the dissemination mechanism. When a new
```

File: docs/core/mempool.md [hunk 3]
```diff
@@ -46,3 +80,24 @@ order/nonce/sequence number, the application can reject transactions that are
 out of order. So if a node receives `tx3`, then `tx1`, it can reject `tx3` and then
 accept `tx1`. The sender can then retry sending `tx3`, which should probably be
 rejected until the node has seen `tx2`.
+
+## 2. Nop
+
+`nop` (short for no operation) mempool is used when the ABCI developer wants to
```
Comment (sergio-mena) [inline docs/core/mempool.md:86]:
```suggestion
`nop` (short for no operation) mempool is used when the ABCI application developer wants to
```

File: docs/core/mempool.md [hunk 4]
```diff
@@ -46,3 +80,24 @@ order/nonce/sequence number, the application can reject transactions that are
...
+`BroadcastTxAsync`, and `BroadcastTxCommit`) will always result in an error.
+
+Because there's no way for the consensus to know if transactions are available
+to be committed, the node will always create blocks, which can be empty
+sometimes. Using `consensus.create_empty_blocks=false` is prohibited in such
+cases.
+
+The ABCI application becomes responsible for storing, disseminating, and
+proposing transactions using [`PrepareProposal`][2]. The concrete design is up
+to the ABCI developers.
```
Comment (sergio-mena) [inline docs/core/mempool.md:100]:
```suggestion
to the ABCI application developers.
```

File: mempool/nop_mempool.go [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+
+// FlushAppConn does nothing.
+func (*NopMempool) FlushAppConn() error { return nil }
+
+// Flush does nothing.
+func (*NopMempool) Flush() {}
+
+// TxsAvailable always returns a closed channel.
+func (*NopMempool) TxsAvailable() <-chan struct{} {
+	txsAvailable := make(chan struct{}, 1)
```
Comment (sergio-mena) [inline mempool/nop_mempool.go:63]: Minor. How often does consensus call this? Would it be worth creating (and closing) a channel upon construction and reusing it?

File: node/node.go [hunk 6]
```diff
@@ -87,6 +87,12 @@ type Node struct {
 	pprofSrv          *http.Server
 }
 
+type mempoolReactor interface {
```
Comment (sergio-mena) [inline node/node.go:90]: Should we define this in the mempool package?

File: node/setup.go [hunk 7]
```diff
@@ -436,7 +444,9 @@ func createSwitch(config *cfg.Config,
 		p2p.SwitchPeerFilters(peerFilters...),
 	)
 	sw.SetLogger(p2pLogger)
-	sw.AddReactor("MEMPOOL", mempoolReactor)
+	if config.Mempool.Type != cfg.MempoolTypeNop {
+		sw.AddReactor("MEMPOOL", mempoolReactor)
```
Comment (sergio-mena) [inline node/setup.go:448]: Will this also prevent declaring the mempool p2p channel(s)?

File: mempool/nop_mempool.go [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+	"github.com/cometbft/cometbft/internal/service"
+	"github.com/cometbft/cometbft/p2p"
+	"github.com/cometbft/cometbft/types"
+)
+
+// NopMempool is a mempool that does nothing.
+//
+// The ABCI app is responsible for storing, disseminating, and proposing transactions.
+// See [ADR-111](../docs/architecture/adr-111-nop-mempool.md).
+type NopMempool struct{}
```
Comment (sergio-mena) [inline mempool/nop_mempool.go:17]: Is there any discrepancy between this implementation and the one described in ADR111? If yes, please adapt the ADR.

**Review by thanethomson — APPROVED** [review]
Just some minor comments, otherwise this looks great, thanks @melekes! :tada:

File: node/setup.go (hunk 7 above)
Comment (melekes) [reply node/setup.go:448]: Yes! Any nodes attempting to send us messages via mempool channels (meaning they have a flood mempool by mistake!) will be disconnected.

File: docs/core/mempool.md [hunk 9]
```diff
@@ -4,7 +4,41 @@ order: 12
...
+A mempool (a contraction of memory and pool) is a node’s data structure for
+storing information on uncommitted transactions. It acts as a sort of waiting
+room for transactions that have not yet been committed.
+
+CometBFT currently supports two types of mempools: `flood` and `nop`.
+
+## 1. Flood
+
+The `flood` mempool stores transactions in a concurrent linked list. When a new
+transaction is received, it first checks there's a space for it (`size` and
```
Comment (lasarojc) [inline docs/core/mempool.md:16]:
```suggestion
transaction is received, it first checks if there's a space for it (`size` and
```

File: docs/core/mempool.md [hunk 10]
```diff
@@ -46,3 +80,24 @@ order/nonce/sequence number, the application can reject transactions that are
 out of order. So if a node receives `tx3`, then `tx1`, it can reject `tx3` and then
 accept `tx1`. The sender can then retry sending `tx3`, which should probably be
 rejected until the node has seen `tx2`.
```
Comment (lasarojc) [inline docs/core/mempool.md:82]: Another approach would be to accept out of order transactions but reorder or exclude them from block proposals, during the execution or `PrepareProposal`

File: docs/core/mempool.md [hunk 11]
```diff
@@ -46,3 +80,24 @@ order/nonce/sequence number, the application can reject transactions that are
...
+
+## 2. Nop
+
+`nop` (short for no operation) mempool is used when the ABCI application developer wants to
+build their own mempool. When `type=nop`, transactions are not stored anywhere
```
Comment (lasarojc) [inline docs/core/mempool.md:87]:
```suggestion
build their own mempool. When `type = "nop"`, transactions are not stored anywhere
```

File: docs/core/mempool.md [hunk 12]
```diff
@@ -46,3 +80,24 @@ order/nonce/sequence number, the application can reject transactions that are
...
+and are not gossiped to other peers using the P2P network.
+
+Submitting a transaction via the existing RPC methods (`BroadcastTxSync`,
+`BroadcastTxAsync`, and `BroadcastTxCommit`) will always result in an error.
+
+Because there's no way for the consensus to know if transactions are available
+to be committed, the node will always create blocks, which can be empty
+sometimes. Using `consensus.create_empty_blocks=false` is prohibited in such
```
Comment (lasarojc) [inline docs/core/mempool.md:95]: Should it be a no-op instead of prohibited?

File: docs/core/configuration.md [hunk 13]
```diff
@@ -305,6 +305,15 @@ dial_timeout = "3s"
...
 [mempool]
 
+# The type of mempool for this node to use.
+#
+#  Possible types:
+#  - "flood" : concurrent linked list mempool with flooding gossip protocol
+#  (default)
+#  - "nop"   : nop-mempool (short for no operation; the ABCI app is
+#  responsible for storing, disseminating and proposing txs)
+type = "flood"
```
Comment (lasarojc) [inline docs/core/configuration.md:315]: extend the comment of `create_empty_blocks`?

File: node/node.go (hunk 6 above)
Comment (lasarojc) [reply node/node.go:90]: Following the service example, maybe `BasicMempoolReactor`

File: mempool/nop_mempool.go [hunk 14]
```diff
@@ -0,0 +1,113 @@
+package mempool
```
Comment (lasarojc) [inline mempool/nop_mempool.go:1]: can't we move each reactor to different subfolder?

File: node/setup.go (hunk 7 above)
Comment (hvanz) [reply node/setup.go:448]: This means that nodes configured by mistake with the flood mempool will not be able to communicate other kind of messages (consensus, sync, etc.) with this node. (I'm ok with that.)

Comment (sergio-mena) [reply node/setup.go:448]: @hvanz I may be wrong, but last time I looked into it, upon handshake, the nodes will transmit/receive only on the intersection of their two channel lists

File: docs/core/mempool.md (hunk 12 above)
Comment (sergio-mena) [reply docs/core/mempool.md:95]: I prefer to have the operator fix their `config.toml`

File: node/setup.go (hunk 7 above)
Comment (melekes) [reply node/setup.go:448]: > Any nodes attempting to send us messages via mempool channels (meaning they have a flood mempool by mistake!) will be disconnected.

_FYI: this statement was checked by running two nodes (one w/ flood mempool and another w/ nop mempool)_

File: config/config.go [hunk 15]
```diff
@@ -894,6 +908,7 @@ type MempoolConfig struct {
 // DefaultMempoolConfig returns a default configuration for the CometBFT mempool
 func DefaultMempoolConfig() *MempoolConfig {
 	return &MempoolConfig{
+		Type:      MempoolTypeFlood,
```
[the remark this answers is not in the document]
Comment (melekes) [reply config/config.go:912]: Done in https://github.com/cometbft/cometbft/pull/1643/commits/3f1ff67bbd63ed1af45f8ee92eadb5c9ab78d11f

File: docs/core/configuration.md (hunk 13 above)
Comment (melekes) [reply docs/core/configuration.md:316]: https://github.com/cometbft/cometbft/pull/1643/commits/f6f3aa6bffac53d86cccc1beec391fc97de9e0a2

File: mempool/nop_mempool.go (hunk 14 above)
Comment (melekes) [reply mempool/nop_mempool.go:1]: I'd prefer to do it in a separate PR.

File: mempool/nop_mempool.go (hunk 8 above)
Comment (melekes) [reply mempool/nop_mempool.go:17]: Done in https://github.com/cometbft/cometbft/pull/1643/commits/4872946bfe6a2ea3756504df1c6924dce2fc5406

File: mempool/nop_mempool.go (hunk 5 above)
Comment (melekes) [reply mempool/nop_mempool.go]: Good point 👍  It's called quite often actually. Changed it to return `nil` https://github.com/cometbft/cometbft/pull/1643/commits/4044579adce425a720569403bf3c713310dd90fa

File: mempool/nop_mempool.go [hunk 16]
```diff
@@ -0,0 +1,113 @@
...
+)
+
+// NopMempool is a mempool that does nothing.
+//
+// The ABCI app is responsible for storing, disseminating, and proposing transactions.
+// See [ADR-111](../docs/architecture/adr-111-nop-mempool.md).
+type NopMempool struct{}
+
+// ErrNotAllowed indicates that the operation is not allowed with `nop` mempool.
+var ErrNotAllowed = errors.New("not allowed with `nop` mempool")
```
[the remark this answers is not in the document]
Comment (melekes) [reply mempool/nop_mempool.go]: Made the error private https://github.com/cometbft/cometbft/pull/1643/commits/84124ae4523aa20ab00196ccfb3631c84718e8c8

**Comment by thanethomson** [comment]
> Is it okay to still open a connection to ABCI app for mempool if it's nop? We could make it optional, but that will require more work.

Looking at the PR, it seems like we'll still open the connection, but we won't ever send anything across, right?

> Should we backport this to 0.34? I personally don't see a need for it right now.

If someone from the community asks for it and has a clear burning need for it. I haven't seen anyone asking for it yet.

**Review by thanethomson — COMMENTED** [review]
One other small thing, I noticed that there's still a ToC entry for the "nil" mempool: https://github.com/cometbft/cometbft/blob/4872946bfe6a2ea3756504df1c6924dce2fc5406/docs/architecture/README.md?plain=1#L48

@melekes would you mind removing that please?

**Comment by melekes** [comment]
> Looking at the PR, it seems like we'll still open the connection, but we won't ever send anything across, right?

Correct

**Comment by sergio-mena** [comment]
1. To me, that's an optimization. If you want to go for it, that's great, but I'd do it in a different PR
2. It makes no sense for `v0.34.x`, as there is no `PrepareProposal` there

**Review by sergio-mena — APPROVED** [review]

**Pull request merged**